File: src/RepositoryModules1/LabelDefinition1.vb
Imports Newtonsoft.Json
Imports Storex

Public Class LabelDefinition1

    Public Shared ReadOnly Property KeyForPrimary As String = "2026aedc-358d-41c5-a6a3-e03abe5ae5b3"

    Public Property PrefixKey As String

    Public Property ItemNumberStartIndex As Integer = 1

    Public Property ItemNumberLength As Integer = 1

    Public Property SerialNumberStartIndex As Integer = 1

    Public Property SerialNumberLength As Integer = 1

    Public Function TryGeneraLabel(Symbol As Symbol, ByRef Label As ILabel) As Boolean

        Dim MinLength As Integer
        MinLength = Math.Max(ItemNumberStartIndex + ItemNumberLength - 1, SerialNumberStartIndex + SerialNumberLength - 1)
        MinLength = Math.Max(MinLength, Len(PrefixKey))

        If Len(Symbol.Value) >= MinLength And (String.IsNullOrEmpty(PrefixKey) OrElse Symbol.Value.StartsWith(PrefixKey)) Then

            Label = New BasicLabel(Symbol) With {
                .ItemNumber = Symbol.Value.Substring(ItemNumberStartIndex - 1, ItemNumberLength).TrimEnd,
                .SerialNumber = Symbol.Value.Substring(SerialNumberStartIndex - 1, SerialNumberLength).TrimEnd
            }
            Return True
        End If

        Return False
    End Function

    Public Shared Function TryExtract(Mode As IMode, Name As String, Optional ByRef Definition As LabelDefinition1 = Nothing) As Boolean

        Dim objValue = Nothing

        objValue = If(Mode.ExtendedProperties.TryGetValue(Name, objValue), objValue, Nothing)

        If TypeOf objValue Is LabelDefinition1 Then

            Definition = objValue

            Return True

        ElseIf objValue IsNot Nothing Then

            Try

                Definition = JsonConvert.DeserializeObject(Of LabelDefinition1)(JsonConvert.SerializeObject(objValue))

                Return True

            Catch ex As JsonException

                Return False
            End Try
        Else

            Return False
        End If
    End Function
End Class

File: src/RepositoryModules.CS/LabelDefinition1.cs
using System;
using Newtonsoft.Json;
using Storex;

namespace RepositoryModules
{
    public class LabelDefinition1
    {
        public static string KeyForPrimary { get; private set; } = "2026aedc-358d-41c5-a6a3-e03abe5ae5b3";

        public string PrefixKey { get; set; }

        public int ItemNumberStartIndex { get; set; } = 1;

        public int ItemNumberLength { get; set; } = 1;

        public int SerialNumberStartIndex { get; set; } = 1;

        public int SerialNumberLength { get; set; } = 1;

        public bool TryGeneraLabel(Symbol symbol, out ILabel label)
        {
            int minLength;
            minLength = Math.Max(ItemNumberStartIndex + ItemNumberLength - 1, SerialNumberStartIndex + SerialNumberLength - 1);
            minLength = Math.Max(minLength, PrefixKey?.Length ?? 0);

            if (symbol.Value.Length >= minLength & (string.IsNullOrEmpty(PrefixKey) || symbol.Value.StartsWith(PrefixKey)))
            {
                label = new BasicLabel(symbol)
                {
                    ItemNumber = symbol.Value.Substring(ItemNumberStartIndex - 1, ItemNumberLength).TrimEnd(),
                    SerialNumber = symbol.Value.Substring(SerialNumberStartIndex - 1, SerialNumberLength).TrimEnd()
                };
                return true;
            }
            else
            {
                label = null;

                return false;
            }
        }

        public static bool TryExtract(IMode mode, string name, out LabelDefinition1 definition)
        {
            if (mode.ExtendedProperties.TryGetValue(name, out var objValue) && objValue is LabelDefinition1 value)
            {
                definition = value;

                return true;
            }
            else if (objValue != null)
            {
                try
                {
                    definition = JsonConvert.DeserializeObject<LabelDefinition1>(JsonConvert.SerializeObject(objValue));

                    return true;
                }
                catch (JsonException)
                {
                    definition = null;

                    return false;
                }
            }
            else
            {
                definition = null;

                return false;
            }
        }
    }
}